### PR TITLE
[Model] Support Qwen2.5-Math-RM-72B

### PR DESCRIPTION
**Pull request by zhuzilin**
FILL IN THE PR DESCRIPTION HERE

Support [Qwen2.5-Math-RM-72B](https://huggingface.co/Qwen/Qwen2.5-Math-RM-72B) as embedding model.

The model can be served with:

```bash
python -m vllm.entrypoints.openai.api_server \
	--model Qwen/Qwen2.5-Math-RM-72B \
	--trust-remote-code \
	--served-model-name Qwen2.5-Math-RM-72B \
	--port 8080 \
	--tensor-parallel-size 8 \
	--enforce-eager
```

And a test correspond to the example in the huggingface model page is:

```python
import torch
from transformers import AutoModel, AutoTokenizer

model_name = "Qwen/Qwen2.5-Math-RM-72B"
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)

chat = [
    {"role": "system", "content": "Please reason step by step, and put your final answer within \\boxed{}."},
    {"role": "user", "content": "Janet’s ducks lay 16 eggs per day. She eats three for breakfast every morning and bakes muffins for her friends every day with four. She sells the remainder at the farmers' market daily for $2 per fresh duck egg. How much in dollars does she make every day at the farmers' market?"},
    {"role": "assistant", "content": "To determine how much Janet makes from selling the duck eggs at the farmers' market, we need to follow these steps:\n\n1. Calculate the total number of eggs laid by the ducks each day.\n2. Determine how many eggs Janet eats and bakes for herself each day.\n3. Find out how many eggs are left to be sold.\n4. Calculate the revenue from selling the remaining eggs at $2 per egg.\n\nLet's start with the first step:\n\n1. Janet's ducks lay 16 eggs per day.\n\nNext, we calculate how many eggs Janet eats and bakes for herself each day:\n\n2. Janet eats 3 eggs for breakfast every morning.\n3. Janet bakes 4 eggs for her friends every day.\n\nSo, the total number of eggs Janet eats and bakes for herself each day is:\n\\[ 3 + 4 = 7 \\text{ eggs} \\]\n\nNow, we find out how many eggs are left to be sold:\n\\[ 16 - 7 = 9 \\text{ eggs} \\]\n\nFinally, we calculate the revenue from selling the remaining eggs at $2 per egg:\n\\[ 9 \\times 2 = 18 \\text{ dollars} \\]\n\nTherefore, Janet makes \\boxed{18} dollars every day at the farmers' market."}
]

conversation_str = tokenizer.apply_chat_template(
    chat, 
    tokenize=False, 
    add_generation_prompt=False
)

from openai import OpenAI

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8080/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

models = client.models.list()
model = models.data[0].id

responses = client.embeddings.create(
    input=[conversation_str],
    model=model,
)

for data in responses.data:
    print(data.embedding[-1])  # 3.3125
```

Currently, the mode will return the score on every token, so that user can use the model as an PRM (process-supervised reward model).

Thank you for your time on reviewing this PR :)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/qwen2_rm.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            ("qkv_proj", "q_proj", "q"),
+            ("qkv_proj", "k_proj", "k"),
+            ("qkv_proj", "v_proj", "v"),
+            ("gate_up_proj", "gate_proj", 0),
+            ("gate_up_proj", "up_proj", 1),
+        ]
+        params_dict = dict(self.named_parameters(remove_duplicate=False))
+        for name, loaded_weight in weights:
+            if name == "lm_head.weight":
+                continue
```
Comment (mgoin) [inline vllm/model_executor/models/qwen2_rm.py:130]: Good to follow the standard and comment this specific case for this model version
```suggestion
            # Skip loading lm_head for embedding model
            if "lm_head" in name:
                continue
```

**Review by mgoin — APPROVED** [review]
LGTM with one nit, thanks for releasing this interesting model to the community. We need more reward models!

To fix the lint issues please `pip install -r requirements-lint.txt` and run `./format.sh`

File: vllm/model_executor/models/qwen2_rm.py (hunk 1 above)
Comment (zhuzilin) [reply vllm/model_executor/models/qwen2_rm.py:133]: fixed.

**Comment by zhuzilin** [comment]
@mgoin I've fixed the lint and add the comment, could you take another look?

**Comment by DarkLight1337** [comment]
Please merge the latest changes from main to fix the CI failures.

Since this model is quite different from what we have currently, I suggest adding an example script for this model under `examples/`.

**Comment by youkaichao** [comment]
we can add it first, and @zhuzilin please add some tests and examples in a followup pr!

reference code: https://github.com/vllm-project/vllm/blob/main/tests/entrypoints/llm/test_encode.py

**Comment by noamgat** [comment]
This is really great! vLLM should definitely have reward models!
A question about the implementation design choice - one downside of this approach is that it requires the client to apply the chat template on their own. Is there a way to address this? (Maybe modify vLLM to accept both a string and a messages list as the parameter to the embedding endpoint? I dunno)

**Comment by zhuzilin** [comment]
@noamgat Agree. I think it would be better if we can introduce a specialize api for reward model, which the openai api does not supports for now. In this way, we can support passing conversations to the reward model, and maybe have a reward score for each response with a PRM. The current implementation is a temporary method to use embedding api to support reward models quickly.

**Comment by wuxianyess** [comment]
现在支持Qwen2.5-Math-RM-72B了么？为什么我下载最新镜像进行部署时报错Model architectures ['Qwen2ForRewardModel'] are not supported for now.

**Comment by DarkLight1337** [comment]
> 现在支持Qwen2.5-Math-RM-72B了么？为什么我下载最新镜像进行部署时报错Model architectures ['Qwen2ForRewardModel'] are not supported for now. 

Yes, but you need to [install vLLM from main branch](https://docs.vllm.ai/en/latest/getting_started/installation.html#build-from-source-with-compilation) since we haven't released a new package version yet.

**Comment by noamgat** [comment]
> @noamgat Agree. I think it would be better if we can introduce a specialized api for reward model, which the openai api does not supports for now. In this way, we can support passing conversations to the reward model, and maybe have a reward score for each response with a PRM. The current implementation is a temporary method to use embedding api to support reward models quickly.

I agree. The vLLM serve API is already a superset of the OpenAI API, with additional options that were introduced here. I was hoping that a core vLLM contributer could chip in - What do you think is the best way to add support for reward modelling in vLLM? Adding messages support to embedding endpoints? Adding "return logits" support to chat endpoints? A completely new endpoint type? @simon-mo  @DarkLight1337 what do you think? Has there been a discussion around this?

**Comment by DarkLight1337** [comment]
> I agree. The vLLM serve API is already a superset of the OpenAI API, with additional options that were introduced here. I was hoping that a core vLLM contributer could chip in - What do you think is the best way to add support for reward modelling in vLLM? Adding messages support to embedding endpoints? Adding "return logits" support to chat endpoints? A completely new endpoint type? @simon-mo @DarkLight1337 what do you think? Has there been a discussion around this?

Not really so far. I suggest you open a RFC to encourage more discussion.

**Comment by simon-mo** [comment]
RFC and suggestion welcomed indeed! I think we are starting to observe the limitation of the OpenAI API.

**Comment by wuxianyess** [comment]
> > 现在支持Qwen2.5-Math-RM-72B了么？为什么我下载最新镜像进行部署时报错Model architectures ['Qwen2ForRewardModel'] are not supported for now.
> 
> Yes, but you need to [install vLLM from main branch](https://docs.vllm.ai/en/latest/getting_started/installation.html#build-from-source-with-compilation) since we haven't released a new package version yet.

Thank you very much, it's done.

**Comment by liziniu** [comment]
Hi, I noticed that in this case, the reward model does not fully utilize the GPU memory, even though there is available free memory. For example, when I use 4 A800-80GB GPU cards with a 72B model, it only takes up 140GB of memory, leaving 180GB unused during the inference stage. Does this seem unusual?

image

**Comment by liziniu** [comment]
Furthermore, I realized that the model takes the same amount of time to process 16 samples simultaneously as it does to process them sequentially with a batch size of 1. Is this normal?

**Comment by DarkLight1337** [comment]
> Hi, I noticed that in this case, the reward model does not fully utilize the GPU memory, even though there is available free memory. For example, when I use 4 A800-80GB GPU cards with a 72B model, it only takes up 140GB of memory, leaving 180GB unused during the inference stage. Does this seem unusual?
> 
> image

Which version of vLLM are you using?

**Comment by liziniu** [comment]
Replace `vllm.model_executor.models.qwen2_rm.py` with the following code, then run `python python_only_dev.py` at the root dir. 

```python

# coding=utf-8
# Adapted from
# https://huggingface.co/Qwen/Qwen2.5-Math-RM-72B/blob/main/modeling_qwen2_rm.py
# Copyright 2024 The Qwen team.
# Copyright 2023 The vLLM team.
"""Inference-only Qwen2-RM model compatible with HuggingFace weights."""
from typing import Iterable, List, Optional, Tuple

import torch
from torch import nn
from transformers import Qwen2Config

from vllm.attention import AttentionMetadata
from vllm.config import CacheConfig, LoRAConfig
from vllm.model_executor.layers.linear import ColumnParallelLinear, RowParallelLinear
from vllm.model_executor.layers.pooler import Pooler, PoolingType
from vllm.model_executor.layers.quantization.base_config import QuantizationConfig
from vllm.model_executor.model_loader.weight_utils import (
    default_weight_loader,
    maybe_remap_kv_scale_name,
)
from vllm.model_executor.models.qwen2 import Qwen2Model
from vllm.model_executor.pooling_metadata import PoolingMetadata
from vllm.sequence import IntermediateTensors, PoolerOutput

from .utils import is_pp_missing_parameter


class ReLU(nn.Module):

    def __init__(self):
        super().__init__()
        self.activation = nn.ReLU()

    def forward(self, input):
        input, _ = input
        return self.activation(input)


class Qwen2ForRewardModel(nn.Module):
    packed_modules_mapping = {
        "qkv_proj": [
            "q_proj",
            "k_proj",
            "v_proj",
        ],
        "gate_up_proj": [
            "gate_proj",
            "up_proj",
        ],
    }

    # LoRA specific attributes
    supported_lora_modules = [
        "qkv_proj",
        "o_proj",
        "gate_up_proj",
        "down_proj",
    ]
    embedding_modules = {}
    embedding_padding_modules = []

    def __init__(
        self,
        config: Qwen2Config,
        cache_config: Optional[CacheConfig] = None,
        quant_config: Optional[QuantizationConfig] = None,
        lora_config: Optional[LoRAConfig] = None,
    ) -> None:
        # TODO (@robertgshaw2): see if this can be moved out
        if cache_config.sliding_window is not None and hasattr(
            config, "max_window_layers"
        ):
            raise ValueError(
                "Sliding window for some but all layers is not "
                "supported. This model uses sliding window "
                "but `max_window_layers` = %s is less than "
                "`num_hidden_layers` = %s. Please open an issue "
                "to discuss this feature."
                % (
                    config.max_window_layers,
                    config.num_hidden_layers,
                )
            )

        super().__init__()

        self.config = config
        self.lora_config = lora_config

        self.quant_config = quant_config
        self.model = Qwen2Model(config, cache_config, quant_config)

        self.score = nn.Sequential(
            ColumnParallelLinear(
                config.hidden_size, config.hidden_size, quant_config=quant_config
            ),
            ReLU(),
            RowParallelLinear(config.hidden_size, 1, quant_config=quant_config),
        )
        self._pooler = Pooler(pooling_type=PoolingType.ALL, normalize=False)

    def forward(
        self,
        input_ids: torch.Tensor,
        positions: torch.Tensor,
        kv_caches: List[torch.Tensor],
        attn_metadata: AttentionMetadata,
        intermediate_tensors: Optional[IntermediateTensors] = None,
    ) -> torch.Tensor:
        hidden_states = self.model(
            input_ids, positions, kv_caches, attn_metadata, intermediate_tensors
        )
        logits, _ = self.score(hidden_states)
        return logits

    def pooler(
        self,
        hidden_states: torch.Tensor,
        pooling_metadata: PoolingMetadata,
    ) -> Optional[PoolerOutput]:
        return self._pooler(hidden_states, pooling_metadata)

    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
        stacked_params_mapping = [
            # (param_name, shard_name, shard_id)
            ("qkv_proj", "q_proj", "q"),
            ("qkv_proj", "k_proj", "k"),
            ("qkv_proj", "v_proj", "v"),
            ("gate_up_proj", "gate_proj", 0),
            ("gate_up_proj", "up_proj", 1),
        ]
        params_dict = dict(self.named_parameters(remove_duplicate=False))
        for name, loaded_weight in weights:
            # Skip loading lm_head for embedding model
            if name == "lm_head.weight":
                continue
            if "rotary_emb.inv_freq" in name:
                continue
            for param_name, weight_name, shard_id in stacked_params_mapping:
                if weight_name not in name:
                    continue
                name = name.replace(weight_name, param_name)
                # Skip loading extra bias for GPTQ models.
                if name.endswith(".bias") and name not in params_dict:
                    continue
                if is_pp_missing_parameter(name, self):
                    continue
                param = params_dict[name]
                weight_loader = param.weight_loader
                weight_loader(param, loaded_weight, shard_id)
                break
            else:
                # Skip loading extra bias for GPTQ models.
                if name.endswith(".bias") and name not in params_dict:
                    continue
                # Remapping the name of FP8 kv-scale.
                name = maybe_remap_kv_scale_name(name, params_dict)
                if name is None:
                    continue
                if is_pp_missing_parameter(name, self):
                    continue
                param = params_dict[name]
                weight_loader = getattr(param, "weight_loader", default_weight_loader)
                weight_loader(param, loaded_weight)

```

**Comment by liziniu** [comment]
> > > > Hi, I noticed that in this case, the reward model does not fully utilize the GPU memory, even though there is available free memory. For example, when I use 4 A800-80GB GPU cards with a 72B model, it only takes up 140GB of memory, leaving 180GB unused during the inference stage. Does this seem unusual?
> > > > image
> > > 
> > > 
> > > Which version of vLLM are you using? I believe the model is not usable in the latest version of vLLM at the moment ([#9090 (comment)](https://github.com/vllm-project/vllm/pull/9090#issuecomment-2418413863)).
> > 
> > 
> > Hi, I am using vllm==0.6.2. I encountered the same issue with ([#9090 (comment)](https://github.com/vllm-project/vllm/pull/9090#issuecomment-2418413863)), and I have fixed the load weight issue in #9090.
> 
> How did you fix it?

Replace `vllm.model_executor.models.qwen2_rm.py` with the following code, then run `python python_only_dev.py` at the root dir. 

```python

# coding=utf-8
# Adapted from
# https://huggingface.co/Qwen/Qwen2.5-Math-RM-72B/blob/main/modeling_qwen2_rm.py
# Copyright 2024 The Qwen team.
# Copyright 2023 The vLLM team.
"""Inference-only Qwen2-RM model compatible with HuggingFace weights."""
from typing import Iterable, List, Optional, Tuple

import torch
from torch import nn
from transformers import Qwen2Config

from vllm.attention import AttentionMetadata
from vllm.config import CacheConfig, LoRAConfig
from vllm.model_executor.layers.linear import ColumnParallelLinear, RowParallelLinear
from vllm.model_executor.layers.pooler import Pooler, PoolingType
from vllm.model_executor.layers.quantization.base_config import QuantizationConfig
from vllm.model_executor.model_loader.weight_utils import (
    default_weight_loader,
    maybe_remap_kv_scale_name,
)
from vllm.model_executor.models.qwen2 import Qwen2Model
from vllm.model_executor.pooling_metadata import PoolingMetadata
from vllm.sequence import IntermediateTensors, PoolerOutput

from .utils import is_pp_missing_parameter


class ReLU(nn.Module):

    def __init__(self):
        super().__init__()
        self.activation = nn.ReLU()

    def forward(self, input):
        input, _ = input
        return self.activation(input)


class Qwen2ForRewardModel(nn.Module):
    packed_modules_mapping = {
        "qkv_proj": [
            "q_proj",
            "k_proj",
            "v_proj",
        ],
        "gate_up_proj": [
            "gate_proj",
            "up_proj",
        ],
    }

    # LoRA specific attributes
    supported_lora_modules = [
        "qkv_proj",
        "o_proj",
        "gate_up_proj",
        "down_proj",
    ]
    embedding_modules = {}
    embedding_padding_modules = []

    def __init__(
        self,
        config: Qwen2Config,
        cache_config: Optional[CacheConfig] = None,
        quant_config: Optional[QuantizationConfig] = None,
        lora_config: Optional[LoRAConfig] = None,
    ) -> None:
        # TODO (@robertgshaw2): see if this can be moved out
        if cache_config.sliding_window is not None and hasattr(
            config, "max_window_layers"
        ):
            raise ValueError(
                "Sliding window for some but all layers is not "
                "supported. This model uses sliding window "
                "but `max_window_layers` = %s is less than "
                "`num_hidden_layers` = %s. Please open an issue "
                "to discuss this feature."
                % (
                    config.max_window_layers,
                    config.num_hidden_layers,
                )
            )

        super().__init__()

        self.config = config
        self.lora_config = lora_config

        self.quant_config = quant_config
        self.model = Qwen2Model(config, cache_config, quant_config)

        self.score = nn.Sequential(
            ColumnParallelLinear(
                config.hidden_size, config.hidden_size, quant_config=quant_config
            ),
            ReLU(),
            RowParallelLinear(config.hidden_size, 1, quant_config=quant_config),
        )
        self._pooler = Pooler(pooling_type=PoolingType.ALL, normalize=False)

    def forward(
        self,
        input_ids: torch.Tensor,
        positions: torch.Tensor,
        kv_caches: List[torch.Tensor],
        attn_metadata: AttentionMetadata,
        intermediate_tensors: Optional[IntermediateTensors] = None,
    ) -> torch.Tensor:
        hidden_states = self.model(
            input_ids, positions, kv_caches, attn_metadata, intermediate_tensors
        )
        logits, _ = self.score(hidden_states)
        return logits

    def pooler(
        self,
        hidden_states: torch.Tensor,
        pooling_metadata: PoolingMetadata,
    ) -> Optional[PoolerOutput]:
        return self._pooler(hidden_states, pooling_metadata)

    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
        stacked_params_mapping = [
            # (param_name, shard_name, shard_id)
            ("qkv_proj", "q_proj", "q"),
            ("qkv_proj", "k_proj", "k"),
            ("qkv_proj", "v_proj", "v"),
            ("gate_up_proj", "gate_proj", 0),
            ("gate_up_proj", "up_proj", 1),
        ]
        params_dict = dict(self.named_parameters(remove_duplicate=False))
        for name, loaded_weight in weights:
            # Skip loading lm_head for embedding model
            if name == "lm_head.weight":
                continue
            if "rotary_emb.inv_freq" in name:
                continue
            for param_name, weight_name, shard_id in stacked_params_mapping:
                if weight_name not in name:
                    continue
                name = name.replace(weight_name, param_name)
                # Skip loading extra bias for GPTQ models.
                if name.endswith(".bias") and name not in params_dict:
                    continue
                if is_pp_missing_parameter(name, self):
                    continue
                param = params_dict[name]
                weight_loader = param.weight_loader
                weight_loader(param, loaded_weight, shard_id)
                break
            else:
                # Skip loading extra bias for GPTQ models.
                if name.endswith(".bias") and name not in params_dict:
                    continue
                # Remapping the name of FP8 kv-scale.
                name = maybe_remap_kv_scale_name(name, params_dict)
                if name is None:
                    continue
                if is_pp_missing_parameter(name, self):
                    continue
                param = params_dict[name]
                weight_loader = getattr(param, "weight_loader", default_weight_loader)
                weight_loader(param, loaded_weight)

```

**Comment by DarkLight1337** [comment]
I see, so we still need to patch that. Any help with https://github.com/vllm-project/vllm/pull/9422#issuecomment-2418872487 would be great!

> Hi, I noticed that in this case, the reward model does not fully utilize the GPU memory, even though there is available free memory. For example, when I use 4 A800-80GB GPU cards with a 72B model, it only takes up 140GB of memory, leaving 180GB unused during the inference stage. Does this seem unusual?

What command are you using to run the model on multiple GPUs?

> Furthermore, I realized that the model takes the same amount of time to process 16 samples simultaneously as it does to process them sequentially with a batch size of 1. Is this normal?

vLLM may wait after the first request to include subsequent requests in the same batch. So, it is possible that the processing time is dominated by the waiting time. That being said, I might be wrong since I don't really work with the optimization side of vLLM - detailed profiling would provide more insights.

**Comment by hulongan** [comment]
do the reward model support offline inference？

**Comment by DarkLight1337** [comment]
Yes, you can use `LLM.encode` in a similar way.

**Pull request merged**